[week2]  Spring Web으로 ToDo REST API 만들기 - kwakcena

Kwakcena opened the pull request:
2주차 과제 제출합니다.

https://user-images.githubusercontent.com/45390172/145197993-f927b050-3b8e-4ae6-88d1-d8b341f1c9ae.mov

Kwakcena committed:
할 일 model을 만들어라

할 일에 대한 model을 구현합니다.

Kwakcena committed:
전체 할 일을 반환하는 api를 구현하라

전체 할 일 목록을 반환하는 api를 만들었습니다.

Kwakcena committed:
할 일을 추가하는 POST api를 만들어라

할 일을 추가하는 POST api를 만들었습니다.

Kwakcena committed:
할 일을 삭제하는 api를 만들어라

할 일 하나를 삭제하는 api를 만들었습니다.
Path에 들어오는 id를 읽기 위해서는 어노테이션에 /{id}를 붙이고 파라메타
입력 부분에는 @PathVariable를 사용하면 된다는 걸 알았습니다.

Kwakcena committed:
spring devtools를 추가합니다

빌드하면 자동으로 서버를 재시작 하는 spring devtools를 추가합니다.

Kwakcena committed:
할 일 목록을 업데이트 하는 api를 만들어라

할 일 목록을 update 하는 patch api를 만들었습니다.

Kwakcena committed:
할일 상세 목록을 반환하는 api를 만들어라

상세 할 일 목록을 가져오는 api를 만들었습니다.

Kwakcena committed:
TaskService를 구현하라

Task에 대한 처리를 수행하는 taskService를 만듭니다.

Kwakcena committed:
CrossOrigin 어노테이션을 추가하라

프론트에서 tasks를 요청할 수 있도록 CrossOrigin 어노테이션을 추가합니다.

Kwakcena committed:
[front] 모든 할 일 목록을 볼 수 있다

프론트에서 모든 할 일 목록을 요청하여 볼 수 있습니다.

Kwakcena committed:
[front] 할 일을 입력하고 추가할 수 있다

사용자가 할 일을 입력하고 추가할 수 있습니다.

Kwakcena committed:
[front] 할 일을 완료할 수 있다

완료 버튼을 누르면 할 일을 제거합니다.

Kwakcena committed:
[front] 할 일을 수정할 수 있다

할 일을 수정할 수 있습니다.

Kwakcena committed:
[front] title component를 구현하라

text를 전달하면 해당 text로 title을 보여주는 컴포넌트를 만들었습니다.

Kwakcena committed:
[front] list component를 만들어라

List를 관리하는 컴포넌트를 만들었습니다.

Kwakcena committed:
[front] input task component를 만들어라

할일 입력창 컴포넌트를 만들었습니다.

Kwakcena committed:
[front] 컴포넌트 로직을 분리하라

List 컴포넌트에서 사용하는 로직을 따로 분리합니다.

Kwakcena committed:
[front] Item component 를 만들어라

하나의 task item 컴포넌트를 만들었습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,49 @@
+// 6. taskService를 만들기 (done)
+
+import com.codesoom.assignment.models.Task;
+import org.springframework.web.bind.annotation.*;
```
`import *`의 사용을 지양하고 어떤 것들이 `import`되는지 분명히 명시해주면 좋을 것 같아요. 가끔 이렇게 `import`한 게 문제가 되는 경우가 있습니다.

Kwakcena replied:
종립님의 블로그 글을 참고하여 설정을 변경하고 사용할 것들을 명시적으로 표현하였습니다.
https://johngrib.github.io/wiki/intellij/#import--%EB%B0%A9%EC%A7%80%ED%95%98%EA%B8%B0

commit: https://github.com/CodeSoom/spring-week2-assignment-1/pull/63/commits/a0c8c54270194e95c9f040489d7199017f35ff1e

johngrib replied:
제 글을 읽어주셔서 감사합니다 :-)

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,49 @@
+@RequestMapping("/tasks")
+@CrossOrigin
+public class TaskController {
+    TaskService taskService = new TaskService();
```
이렇게 선언한 `TaskService`는 `TaskController` 클래스 바깥에서 참조할 수 없도록 접근 제한을 최소로 지정해 주면 어떨까요?
Java로 코딩할 때에는 가능한 한 많은 멤버를 `private`으로 선언해서 외부의 접근을 최대한 차단하고, 공개된 메소드를 통해서만 상태 변화를 허용하도록 작업해 보세요.

```suggestion
    private TaskService taskService = new TaskService();
```

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,44 @@
+        return task;
+    }
+
+    public Task updateTask(Long id, Task source) {
+        Task task = getTask(id);
+        task.setTitle(source.getTitle());
```
`getTask`가 `null`을 리턴할 가능성이 있기 때문에, `task.setTitle`을 호출할 때 Null Pointer Exception이 발생할 위험이 있습니다. 이 문제를 해결하기 위해 `Optional`을 사용하는 것을 권해드려요. 가급적이면 `null`을 사용하지 않는 것이 좋습니다. 

https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/Optional.html
https://johngrib.github.io/wiki/java-optional/

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,44 @@
+        tasks.remove(task);
+    }
+
+    private Long generateId() {
+        newId += 1;
+        return newId;
+    }
```
Java의 `synchronized` 키워드에 대해 조사해보시고 여기에 적용할 필요가 있을지 생각해보세요.

Kwakcena replied:
`synchronized` 키워드가 무엇인지, 이 상황에서 왜 사용해야 하는지에 대해서 적어보았습니다. 그리고 https://github.com/CodeSoom/spring-week2-assignment-1/pull/63/commits/b6e051cba0c7f9c0a538a3f545f9eed47d385276 에서 이를 적용했습니다.
- https://diamond-shamrock-da0.notion.site/synchronized-07dd223c71594a46b2b896fe71e93bf1

johngrib replied:
오 너무 좋은 글 같아요. 읽으면서 무척 재미있었습니다.

johngrib replied:
시간이 충분하실 때 JLS도 함께 읽어보시길 권해드려요. https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html

Kwakcena committed:
import * 의 사용을 지양하라

import * 를 하지 않고, 내가 사용할 것들을 명시적으로 보여준다.

Kwakcena committed:
TaskService의 접근 제한자를 private로 설정한다

| 자바에서 코딩할 때는 가능한 한 많은 멤버를 private으로 선언해서 외부의
접근을 최대한 차단하고, 공개된 메소드를 통해서만 상태 변화를 허용하도록
작업 해 보세요

Kwakcena committed:
Synchronized 키워드를 적용하라

동시에 들어오는 요청이 있을 경우 공유 변수가 손상될 우려가 있다면
인스턴스 메소드에 synchronized 키워드를 적용하여 이를 방지할 수 있다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,57 @@
+//  -> 없을 경우 not found 처리
+// 3. POST /tasks (done)
+// 4. PUT/PATCH /tasks/{id} (done)
+//  -> PUT은 어떻게 하지? PUT과 PATCH mapping을 같이 쓸 순 없나?
```
이렇게 하시면 여러 개의 HTTP method에 대응하게 할 수 있어요.
```java
@RequestMapping(value = "/tasks/{id}", method = {RequestMethod.PUT, RequestMethod.PATCH})
```

Kwakcena replied:
오 이런 방법이 ㅎㅎ 감사합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TaskService {
```
지금은 `TaskController`에서 `new TaskService`해서 사용하고 있지만, `TaskService`를 여러 곳에서 사용할 수도 있겠죠.
여유가 있으실 때 이 클래스를 `@Service`나 `@Component`로 선언해서 스프링이 관리하게 작업해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,44 @@
+        return tasks;
+    }
+
+    public Task getTask(Long id) {
+        return tasks.stream()
+                .filter(task -> task.getId().equals(id))
+                .findFirst()
+                .orElse(null);
+    }
```
`Task`를 찾는 작업이 `O(n)`인 점을 개선해 보는 것이 좋겠어요. `id`가 작을수록 빨리 찾겠지만, `id`는 서비스를 운영하면서 점점 커지는 숫자가 되겠죠.

Kwakcena committed:
update 메소드에 PUT, PATCH를 같이 할당한다

RequestMapping을 이용하면 여러개의 메소드를 매칭시킬 수 있다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,57 @@
+        return taskService.createTask(task);
+    }
+
+    @RequestMapping(value = "/{id}", method = {RequestMethod.PUT, RequestMethod.PATCH})
```
`@PutMapping`과 `@PatchMapping`의 소스코드도 읽어보시면 도움이 될 거에요. 들여다보면 `@RequestMapping`을 사용하고 있다는 사실을 알 수 있어요.

![image](https://user-images.githubusercontent.com/1855714/145571995-3077db10-9983-4b74-986e-ec2b51eea2c4.png)

johngrib reviewed (commented):
한 주간 고생 많으셨어요. 어떠셨나요? 너무 쉽거나 너무 어렵지는 않으셨나요? 주간 회고를 작성해 공유해 주세요.

Kwakcena committed:
[풀이] ResponseEntity를 이용하여 없을 경우 404 반환

Optional을 이용한 ResponseEntity를 사용하여 없을 경우에는 상태코드 404를 반환하도록 합니다.

Kwakcena committed:
[풀이] service 에서 ResponseEntity를 반환하게 하라

taskService가 ResponseEntity를 반환하도록 하면 실제 컨트롤러에서는 코드를 깔끔하게 만들 수 있는것 같다.

Kwakcena committed:
[풀이] orElseThrow 사용, error dto 생성

task를 찾을 때 에러가 발생하면 이에 대한 에러를 처리해주는 TaskErrorAdvice 클래스를 만들었다.
TaskErrorAdvice는 아무런 task도 찾지 못했을 때에 대한 handler가 등록되어 있고, 이러한 advice는
클래스 위에 @controllerAdvice가 추가된다.

응답값을 마음대로 수정하기 위해서는 @ResponseBody를 붙여준다.

Kwakcena committed:
[풀이] TaskService의 사용, marko 수정

TaskService를 사용하고 marko를 수정합니다.
